docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,668 @@
+          /\ HandleAppendEntriesRequest(i, j, m)
+       \/ /\ m.mtype = AppendEntriesResponse
+          /\ \/ DropStaleResponse(i, j, m)
+             \/ HandleAppendEntriesResponse(i, j, m)
+    ...
+Next ==
+    ...
+           \/ \E m \in DOMAIN messages : Receive(m)
+           \/ \E m \in DOMAIN messages : DuplicateMessage(m)
+           \/ \E m \in DOMAIN messages : DropMessage(m)
+    ...
+```
+
+### 3. Other issues
+
+As can be seen from the Paxos example, we should take care of sets of
+records that are used as sets of variants:
+
+```tla
+Message ==      [type : {"1a"}, bal : Ballot]
+           \cup [type : {"1b"}, acc : Acceptor, bal : Ballot, 
+                 mbal : Ballot \cup {-1}, mval : Value \cup {None}]
+           \cup [type : {"2a"}, bal : Ballot, val : Value]
+           \cup [type : {"2b"}, acc : Acceptor, bal : Ballot, val : Value]
+```           
+
+## 4. Options
+
+There are several solutions to the issue of precise record typing.
+
+### 4.1. Support only plain records
+
+In this case, we would only allow to mix records that have exactly the same
+shape. As a result, when we use variants, we would have to add extra
+fields to all of them. This solution is not very different from the current
+implementation of record type checking, though it would allow us to quickly
+detect spelling errors.
+
+**Blocker.** This does not look like a real solution, as it would immediately
+render the existing examples invalid. Moreover, they would be no obvious way to
+repair these examples.
+
+### 4.2. Support plain records and variants, but no row typing
+
+In this scenario, the type checker would issue an error, if a record expression
+accesses a field that is outside of its type:
+
+```tla
+FieldAccess ==
+  LET m == [ a |-> 2, b |-> "B" ] IN
+  /\ m.a > 1        \* type OK
+  /\ m.b = "B"      \* type OK
+  /\ m.c = { 1, 2 } \* should flag a type error
+```
+
+Moreover, the following example would require a type annotation:
+
+```tla
+RowAccess(m) ==
+    m.a > 0         \* should flag a type error
+```
+
+In the above example, the type checker would not be able to infer the type of
+`m` and would require an explicit type annotation:
+
+
+```tla
+\* @type: [ a: Int, b: Str ];
+RowAccess(m) ==
+    m.a > 0         \* should not flag a type error
+```
+
+**Blocker.** It is not clear to me, what type we would assign to the record
+access operator. Although the type of `m` in `FieldAccess` is obvious to a
+human reader, as it simply requires us to do the top-to-the-bottom type
+propagation, Snowcat constructs a set of equality constraints that are solved
+by unification. This approach requires that we capture the record access
+operator `.` as an equality over type variables and types.
+
+### 4.3. Support plain records and variants, including row typing
+
+This solution follows the approach outlined by [Leijen05][], except that we
+require field names to be unique, not scoped. We present our approach in the
+next section.
```

Comment:
As per my comment above -- that clearly stating the many aspects of the paper are not relevant here would help with clarity -- what about extending this with something like the following:

```suggestion
This solution is inspired by the approach outlined by [Leijen05][], but is much more limited. The following features discussed in [Leijen05][] are neither needed needed nor supported:

- extension for records
- extension for variants
- record restriction
- scoped labels

Our use is limited to a few special cases that give support for subtyping and incremental inference of anonymous record types, in which we cannot know the full set of fields up front. 
```